### PR TITLE
Introduce Filesystem Finder and Path utilities

### PR DESCRIPTION
**Pull request by josbeir**
**Refs #19167**

This PR introduces a new `Cake\Utility\Fs` namespace with filesystem utilities:

## 1. Path - Static Path Manipulation Utilities

Cross-platform utilities for working with file paths:

```php
use Cake\Utility\Fs\Path;

// Normalize paths to use forward slashes
Path::normalize('path\\to\\file.php');
// Returns: 'path/to/file.php'

// Make paths relative
Path::makeRelative('/var/www/app/src/Controller/UsersController.php', '/var/www/app');
// Returns: 'src/Controller/UsersController.php'

// Join path segments
Path::join('src', 'Controller', 'UsersController.php');
// Returns: 'src/Controller/UsersController.php'

// Match glob patterns
Path::matches('*.php', 'test.php');        // true
Path::matches('src/**/*.php', 'src/Controller/UsersController.php');  // true
```

## 2. Finder - Fluent File and Directory Discovery

Lazy iterator-based finder with fluent interface for discovering files and directories:

### Basic Usage

```php
use Cake\Utility\Fs\Finder;

// Find files
$finder = new Finder();
$files = $finder
    ->in('src')
    ->name('*.php')
    ->files();

foreach ($files as $file) {
    echo $file->getPathname();
}

// Find directories
$directories = (new Finder())
    ->in('src')
    ->exclude('vendor')
    ->directories();

// Find both files and directories
$all = (new Finder())
    ->in('src')
    ->all();
```

### Recursive Control

```php
// Non-recursive search (top-level only)
$finder = (new Finder())
    ->in('src')
    ->recursive(false)
    ->files();

// Recursive search (default)
$finder = (new Finder())
    ->in('src')
    ->recursive(true)  // Can be omitted, true by default
    ->files();
```

### Filter by Filename Patterns

```php
// Include specific patterns
$finder = (new Finder())
    ->in('src')
    ->name('*Controller.php')
    ->name('*Table.php')
    ->files();

// Exclude specific patterns
$finder = (new Finder())
    ->in('src')
    ->name('*.php')
    ->notName('*Test.php')
    ->notName('*Fixture.php')
    ->files();

// Combine include and exclude
$finder = (new Finder())
    ->in('src')
    ->name('User*.php')          // Match User*.php
    ->notName('*Table.php')      // But exclude *Table.php
    ->files();
```

### Exclude Directories

```php
$finder = (new Finder())
    ->in('.')
    ->exclude('vendor')
    ->exclude('tests')
    ->exclude('tmp')
    ->files();
```

### Filter by Path Patterns

```php
// Include paths containing pattern
$finder = (new Finder())
    ->in('src')
    ->path('Controller')
    ->files();

// Exclude paths containing pattern
$finder = (new Finder())
    ->in('src')
    ->notPath('vendor')
    ->notPath('Test')
    ->files();

// Use regex patterns for advanced matching
$finder = (new Finder())
    ->in('src')
    ->path('/Controller\.php$/')  // Regex pattern
    ->files();
```

### Glob Pattern Matching

```php
// Recursive glob patterns
$finder = (new Finder())
    ->in('.')
    ->pattern('src/**/*Controller.php')
    ->pattern('tests/**/*Test.php')
    ->files();
```

### Custom Filters

For edge cases not covered by built-in filters, use custom callbacks:

```php
use SplFileInfo;

// Filter by file size
$finder = (new Finder())
    ->in('src')
    ->filter(fn(SplFileInfo $file) => $file->getSize() > 1024)
    ->files();

// Filter using relative path (second parameter)
$finder = (new Finder())
    ->in('.')
    ->filter(function (SplFileInfo $file, string $relativePath) {
        // Only files in specific subdirectories
        return str_starts_with($relativePath, 'src/Controller') 
            || str_starts_with($relativePath, 'src/Model');
    })
    ->files();

// Combine multiple custom filters (AND logic)
$finder = (new Finder())
    ->in('src')
    ->filter(fn(SplFileInfo $file) => $file->getSize() > 1024)
    ->filter(fn(SplFileInfo $file) => $file->getMTime() > strtotime('-1 week'))
    ->files();
```

**Note:** Custom filters are applied as the final filtering stage and should only be used when built-in filters don't meet your needs.

### Depth Control (Type-Safe with Enum)

```php
use Cake\Utility\Fs\Enum\DepthOperator;

// Only top-level files (automatically optimized)
$finder = (new Finder())
    ->in('src')
    ->depth(0)           // Defaults to DepthOperator::EQUAL
    ->files();

// Or use recursive(false) for the same behavior
$finder = (new Finder())
    ->in('src')
    ->recursive(false)   // Internally uses depth=0
    ->files();

// Maximum depth of 3
$finder = (new Finder())
    ->in('src')
    ->depth(3, DepthOperator::LESS_THAN)
    ->files();

// Range: depth between 1 and 3 (multiple conditions use AND logic)
$finder = (new Finder())
    ->in('src')
    ->depth(0, DepthOperator::GREATER_THAN)      // Greater than 0
    ->depth(4, DepthOperator::LESS_THAN)         // Less than 4
    ->files();

// Available operators (DepthOperator enum):
// - EQUAL (==)
// - NOT_EQUAL (!=)
// - LESS_THAN (<)
// - GREATER_THAN (>)
// - LESS_THAN_OR_EQUAL (<=)
// - GREATER_THAN_OR_EQUAL (>=)

$finder = (new Finder())
    ->in('src')
    ->depth(2, DepthOperator::LESS_THAN_OR_EQUAL)     // Depth 0, 1, or 2
    ->files();
```

### Finding Directories

```php
// Find all directories
$directories = (new Finder())
    ->in('src')
    ->directories();

// Find directories matching pattern
$directories = (new Finder())
    ->in('src')
    ->name('*Controller')
    ->directories();

// Find top-level directories only
$directories = (new Finder())
    ->in('src')
    ->recursive(false)
    ->directories();

// Exclude specific directories
$directories = (new Finder())
    ->in('src')
    ->exclude('vendor')
    ->exclude('tmp')
    ->directories();
```

### Complex Example

```php
use Cake\Utility\Fs\Enum\DepthOperator;

$finder = (new Finder())
    ->in('src')
    ->in('plugins')
    ->name('*.php')
    ->notName('*Test.php')
    ->notName('*Fixture.php')
    ->exclude('vendor')
    ->exclude('tmp')
    ->path('Controller')
    ->notPath('Test')
    ->depth(5, DepthOperator::LESS_THAN)
    ->ignoreHiddenFiles()
    ->files();

$count = 0;
foreach ($finder as $file) {
    echo $file->getRealPath() . PHP_EOL;
    $count++;
}
echo "Found {$count} files\n";
```

## Benefits

- **Lazy evaluation**: Files are only iterated when needed, efficient for large directories
- **Fluent interface**: Chainable methods for readable queries
- **Cross-platform**: Works consistently on Windows/Linux/Mac
- **Memory efficient**: Iterator-based traversal, no buffering unless explicitly needed
- **Type-safe**: Enums for modes (FILES, DIRECTORIES, ALL) and depth operators
- **Flexible**: Find files, directories, or both with the same API

### Iterator Architecture

All filtering is implemented through **dedicated iterator classes** following clean design principles:

**8 Custom Iterator Classes:**

1. **HiddenFileFilterIterator** - Filters hidden files/directories (starting with '.')
2. **ExcludeDirectoryFilterIterator** - Excludes directories by name during recursive traversal
3. **ContainsPathFilterIterator** - Includes/excludes files by path patterns (supports both substring matching and regex patterns, with negation support)
4. **FileTypeFilterIterator** - Filters by type (FILES/DIRECTORIES/ALL mode)
5. **FilenameFilterIterator** - Glob-based filename matching with include/exclude support via negation parameter
6. **DepthFilterIterator** - Depth-based filtering with type-safe operators
7. **GlobFilterIterator** - Glob pattern matching for relative paths (for `pattern()`)
8. **CallbackFilterIterator** - Custom closure-based filtering (for `filter()`)

## Refactored Existing Code

This PR also migrates existing CakePHP code to use the new Finder API:

### CommandScanner.php
**Before:**
```php
$fs = new Filesystem();
$files = $fs->find($path, '/Command\.php$/');
```

**After:**
```php
use Cake\Utility\Fs\Finder;

$files = (new Finder())
    ->in($path)
    ->recursive(false)
    ->name('*Command.php')
    ->files();
```

### I18nExtractCommand.php
**Before:**
```php
$fs = new Filesystem();
$files = $fs->findRecursive($path, '/\.php$/');
$files = array_keys(iterator_to_array($files));
sort($files);
```

**After:**
```php
use Cake\Utility\Fs\Finder;

$files = (new Finder())
    ->in($path)
    ->name('*.php')
    ->files();
foreach ($files as $file) {
    $this->_files[] = $file->getPathname();
}
sort($this->_files);
```

Both refactorings maintain backward compatibility while demonstrating cleaner, more expressive code.

## Backward Compatibility

- `Utility\Filesystem::find()` and `findRecursive()` remain BC
- Finder is a new addition with no BC concerns

## Testing

- 94 tests with 323 assertions covering all Finder functionality
- All iterator classes have comprehensive unit tests
- Tests verify pattern matching, negation, recursion, depth control, and filter combinations
- PHPCS and PHPStan level 8 compliant

### CODE REVIEW

File: src/Filesystem/Path.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            return '';
+        }
+
+        $result = array_shift($segments);
+
+        foreach ($segments as $segment) {
+            $result = rtrim($result, '/\\') . '/' . ltrim($segment, '/\\');
+        }
+
+        return static::normalize($result);
```
Comment (ADmad) [inline src/Filesystem/Path.php]: Is the call to normalize() necessary when you are already trimming both type of slashes and then joining with `/`?

Comment (josbeir) [reply src/Filesystem/Path.php]: It was to ensure that windows style segments are properly normalized but refactored now to handle it a bit differently.

File: src/Filesystem/Finder.php [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+    {
+        $this->globPatterns[] = $pattern;
+
+        return $this;
+    }
+
+    /**
+     * Add a depth condition.
+     *
+     * @param string $condition Depth condition (e.g., '== 0', '< 3')
```
Comment (ADmad) [inline src/Filesystem/Finder.php]: I dislike args which require string parsing. Can't we have separate arguments for the operator and the number?

File: src/Filesystem/Finder.php [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+     * @return $this
+     */
+    public function ignoreHiddenFiles(bool $ignore = true)
+    {
+        $this->ignoreHiddenFiles = $ignore;
+
+        return $this;
+    }
+
+    /**     * Set whether to search recursively into subdirectories.
```
Comment (ADmad) [inline src/Filesystem/Finder.php]: Formatting

File: src/Filesystem/Finder.php (hunk 2 above)
Comment (josbeir) [reply src/Filesystem/Finder.php]: Yeah, updated using $depth and $operator arguments

File: src/Utility/Fs/Path.php [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+{
+    /**
+     * Normalizes a path by converting directory separators to forward slashes
+     * and removing redundant separators.
+     *
+     * @param string $path The path to normalize
+     * @param bool $trailing Whether to preserve trailing slashes
+     * @return string The normalized path
+     */
+    public static function normalize(string $path, bool $trailing = false): string
```
Comment (othercorey) [inline src/Utility/Fs/Path.php:35]: We have pathCombine() in Core. Should this be moved there or that moved here?

Comment (othercorey) [reply src/Utility/Fs/Path.php:35]: If we move it out of Core, then should all of this just go into Core? Otherwise you just have to add a dependency on Filesystem.

Is there any real need for this to be separate?

Comment (josbeir) [reply src/Utility/Fs/Path.php:35]: I refactored Path::join to make it compatible with that method but with variadic arguments and a more performant implementation.

I think a deprecation of `pathCombine` would be a good idea. I didn't do it just yet because i don't want to tie this PR too much into current core but i can add it everybody is in agreement of this?

*update*: reverted as per @ADmad 's comment

File: src/Filesystem/Path.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    /**
+     * Joins path segments together.
+     *
+     * Preserves existing directory separators (/ or \) from the input segments.
+     * Use Path::normalize() if you need to convert all separators to forward slashes.
+     *
+     * @param string|bool|null ...$segments Path segments to join. The last argument
+     *  can be a bool|null to control trailing slash handling:
```
Comment (ADmad) [inline src/Filesystem/Path.php]: This is ugly, widening the variadic arg type means one can potentially pass non string segments too. It also unnecessarily complicates the code.

I would rather we forget compatibility with `pathCombine()`. I don't think its a huge problem if both exist.

Comment (josbeir) [reply src/Filesystem/Path.php]: Hmm 🔨 

You have a point.

I also noticed it is not used in core so it can eventually be removed at a later time probably?

**Review by ADmad — APPROVED** [review]

File: src/Utility/Fs/Path.php (hunk 4 above)
Comment (othercorey) [reply src/Utility/Fs/Path.php:35]: My point was deprecating the Core function then requires a dependency on Filesystem so is there any real need to keep it separate?

**Comment by othercorey** [comment]
My comment got closed so I'm re-posting here.

I'm not sure we need a separate Filesystem package/namespace.. Why can't this go into Core? We have pathCombine() and moving that would put a dependency on a different package. Keeping it separate feels odd as well.

**Comment by ADmad** [comment]
> I'm not sure we need a separate Filesystem package/namespace.. Why can't this go into Core? We have pathCombine() and moving that would put a dependency on a different package. Keeping it separate feels odd as well.

I have an opposing opinion. I don't think file finder is a core functionality. The `Core` package is used by every split package and including the files related classes in Core doesn't feel right to me. `pathCombine()` is a single function and hence got put into `Core/functions.php`.  In the past too we have a separate `Filesystem` namespace (which got dropped as its `File` and `Folder` classes were quite quirky and bad alternative to simply using `FileInfo`).  `Utility/Filesystem` was kept as an internal class to keep open the possibility of having a new/better API under the `Filesystem` namespace.

**Comment by josbeir** [comment]
Changes:

* I refactored the namespace from `Filesystem` to `Utility/Fs` as discussed on chat
* Updated using custom Iterator classes. Finder is now only responsible for building an iterator. All other logic is delegated to the iterator classes.

**Review by othercorey — APPROVED** [review]

**Review by ADmad — APPROVED** [review]

**Review by LordSimal — APPROVED** [review]

**Comment by othercorey** [comment]
@josbeir Can you add a migration note for the new classes?

**Comment by josbeir** [comment]
> @josbeir Can you add a migration note for the new classes?

done in cakephp/docs#8154
